src/Data/Variant.hs

```diff
@@ -276,6 +279,36 @@ class CouldBeF xs x => CouldBe (xs :: [Type]) (x :: Type) where
 instance CouldBeF xs x => CouldBe xs x where
   throw = throwF . Identity
 
+type family FlattenConstraints (cs :: [Constraint]) = (c :: Constraint) | c -> cs where
+  FlattenConstraints  '[] = ()
+  FlattenConstraints (c ': cs) = (c, FlattenConstraints cs)
+
+type family TypeMap (f :: k -> l) (xs :: [k]) = (c :: [l]) where
+  TypeMap f (x ': xs) = f x ': (TypeMap f xs)
+  TypeMap f '[] = '[]
+
+type family SpreadConstraint (f :: k -> Constraint) (xs :: [k]) = (c :: Constraint) where
```

Comment:
With the sweet machinery you defined above, we can make these definitions much less terrifying and drop `SpreadConstraint` entirely:

```
type e `CouldBeAnyOf`  xs = All (Map (e `CouldBe`) xs)
type e `CouldBeAnyOfF` xs = All (Map (e `CouldBeF`) xs)```